MoveThreadCount + kOptListMoveSelector throws IndexOutOfBoundsException

Reproducible with the following solver configuration:
```
<?xml version="1.0" encoding="UTF-8"?>
<solver xmlns="https://timefold.ai/xsd/solver" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="https://timefold.ai/xsd/solver https://timefold.ai/xsd/solver/solver.xsd">
    <moveThreadCount>4</moveThreadCount>
    <constructionHeuristic/>
    <localSearch>
        <unionMoveSelector>
            <listChangeMoveSelector/>
            <listSwapMoveSelector/>
            <kOptListMoveSelector>
                <maximumK>3</maximumK>
            </kOptListMoveSelector>
        </unionMoveSelector>
    </localSearch>
</solver>
```

Not reproducible when using moveThreadCount = NONE.
Reproducible with moveThreadCount = 1.

Exception stacktrace:
```
13:31:20.654 ERROR [pool-9-thread-1] Solving with id (78fc989b-d863-4727-b509-62e33e5fe0b8) failed: message (The move thread with moveThreadIndex (0) has thrown an exception. Relayed here in the parent thread.): java.lang.IllegalStateException: The move thread with moveThreadIndex (0) has thrown an exception. Relayed here in the parent thread.
	at ai.timefold.solver.enterprise.core.multithreaded.OrderByMoveIndexBlockingQueue.take(OrderByMoveIndexBlockingQueue.java:111)
	at ai.timefold.solver.enterprise.core.multithreaded.MultiThreadedLocalSearchDecider.forageResult(MultiThreadedLocalSearchDecider.java:180)
	at ai.timefold.solver.enterprise.core.multithreaded.MultiThreadedLocalSearchDecider.decideNextStep(MultiThreadedLocalSearchDecider.java:144)
	at ai.timefold.solver.core.impl.localsearch.DefaultLocalSearchPhase.solve(DefaultLocalSearchPhase.java:72)
	at ai.timefold.solver.core.impl.solver.AbstractSolver.runPhases(AbstractSolver.java:83)
	at ai.timefold.solver.core.impl.solver.DefaultSolver.solve(DefaultSolver.java:193)
	at ai.timefold.solver.core.impl.solver.DefaultSolverJob.call(DefaultSolverJob.java:110)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IndexOutOfBoundsException: toIndex = 9
	at java.base/java.util.AbstractList.subListRangeCheck(AbstractList.java:507)
	at java.base/java.util.ArrayList.subList(ArrayList.java:1108)
	at ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.MultipleDelegateList.subList(MultipleDelegateList.java:271)
	at ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.MultipleDelegateList.moveElementsOfDelegates(MultipleDelegateList.java:86)
	at ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.KOptListMove$UndoKOptListMove.doMoveOnGenuineVariables(KOptListMove.java:211)
	at ai.timefold.solver.core.impl.heuristic.move.AbstractMove.doMoveOnly(AbstractMove.java:26)
	at ai.timefold.solver.core.impl.score.director.AbstractScoreDirector.doAndProcessMove(AbstractScoreDirector.java:197)
	at ai.timefold.solver.enterprise.core.multithreaded.MoveThreadRunner.run(MoveThreadRunner.java:135)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	... 4 more
```

--- a/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/FlipSublistAction.java
+++ b/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/FlipSublistAction.java
@@ -4,9 +4,6 @@ import java.util.Collections;
 import java.util.List;
 
 import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescriptor;
-import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
-import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonListInverseVariableDemand;
-import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
 
 /**
  * Flips a sublist of a list variable, (the same thing as a {@link TwoOptListMove}, but no shift to restore the original
@@ -24,23 +21,20 @@ final class FlipSublistAction {
 
     private final MultipleDelegateList<Object> combinedList;
 
-    private final SingletonInverseVariableSupply inverseVariableSupply;
     private final int fromIndexInclusive;
     private final int toIndexExclusive;
 
     public FlipSublistAction(ListVariableDescriptor<?> variableDescriptor,
-            SingletonInverseVariableSupply inverseVariableSupply,
             MultipleDelegateList<Object> combinedList,
             int fromIndexInclusive, int toIndexExclusive) {
         this.variableDescriptor = variableDescriptor;
-        this.inverseVariableSupply = inverseVariableSupply;
         this.combinedList = combinedList;
         this.fromIndexInclusive = fromIndexInclusive;
         this.toIndexExclusive = toIndexExclusive;
     }
 
     FlipSublistAction createUndoMove() {
-        return new FlipSublistAction(variableDescriptor, inverseVariableSupply,
+        return new FlipSublistAction(variableDescriptor,
                 combinedList,
                 fromIndexInclusive,
                 toIndexExclusive);
@@ -62,12 +56,9 @@ final class FlipSublistAction {
         flipSublist(combinedList, fromIndexInclusive, toIndexExclusive);
     }
 
-    public FlipSublistAction rebase(InnerScoreDirector<?, ?> destinationScoreDirector) {
-        SingletonInverseVariableSupply newInverseVariableSupply = destinationScoreDirector.getSupplyManager()
-                .demand(new SingletonListInverseVariableDemand<>(variableDescriptor));
+    public FlipSublistAction rebase(MultipleDelegateList<Object> rebasedList) {
         return new FlipSublistAction(variableDescriptor,
-                newInverseVariableSupply,
-                combinedList.rebase(variableDescriptor, inverseVariableSupply, destinationScoreDirector),
+                rebasedList,
                 fromIndexInclusive,
                 toIndexExclusive);
     }

--- a/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptDescriptor.java
+++ b/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptDescriptor.java
@@ -254,7 +254,7 @@ final class KOptDescriptor<Node_> {
             }
             if (maximumOrientedPairCountAfterReversal >= 0) {
                 if ((bestOrientedPairFirstEndpoint & 1) == 1) {
-                    out.add(getListReversalMoveForEdgePair(listVariableDescriptor, inverseVariableSupply, indexVariableSupply,
+                    out.add(getListReversalMoveForEdgePair(listVariableDescriptor, indexVariableSupply,
                             combinedList, originalToCurrentIndexList,
                             removedEdges[currentRemovedEdgeIndexToTourOrder[bestOrientedPairFirstEndpoint + 1]],
                             removedEdges[currentRemovedEdgeIndexToTourOrder[bestOrientedPairFirstEndpoint]],
@@ -265,7 +265,7 @@ final class KOptDescriptor<Node_> {
                             bestOrientedPairFirstEndpoint + 1,
                             bestOrientedPairSecondEndpoint);
                 } else {
-                    out.add(getListReversalMoveForEdgePair(listVariableDescriptor, inverseVariableSupply, indexVariableSupply,
+                    out.add(getListReversalMoveForEdgePair(listVariableDescriptor, indexVariableSupply,
                             combinedList, originalToCurrentIndexList,
                             removedEdges[currentRemovedEdgeIndexToTourOrder[bestOrientedPairFirstEndpoint - 1]],
                             removedEdges[currentRemovedEdgeIndexToTourOrder[bestOrientedPairFirstEndpoint]],
@@ -286,7 +286,7 @@ final class KOptDescriptor<Node_> {
                 int secondEndpoint = currentInverseRemovedEdgeIndexToTourOrder[nextEndpointTourIndex];
 
                 if (secondEndpoint >= firstEndpoint + 2) {
-                    out.add(getListReversalMoveForEdgePair(listVariableDescriptor, inverseVariableSupply, indexVariableSupply,
+                    out.add(getListReversalMoveForEdgePair(listVariableDescriptor, indexVariableSupply,
                             combinedList, originalToCurrentIndexList,
                             removedEdges[currentRemovedEdgeIndexToTourOrder[firstEndpoint]],
                             removedEdges[currentRemovedEdgeIndexToTourOrder[firstEndpoint + 1]],
@@ -431,7 +431,6 @@ final class KOptDescriptor<Node_> {
     @SuppressWarnings("unchecked")
     private static <Node_> FlipSublistAction getListReversalMoveForEdgePair(
             ListVariableDescriptor<?> listVariableDescriptor,
-            SingletonInverseVariableSupply inverseVariableSupply,
             IndexVariableSupply indexVariableSupply,
             MultipleDelegateList<Node_> combinedList,
             int[] originalToCurrentIndexList,
@@ -455,7 +454,7 @@ final class KOptDescriptor<Node_> {
 
         KOptUtils.flipSubarray(originalToCurrentIndexList, firstEndpoint, secondEndpoint);
 
-        return new FlipSublistAction(listVariableDescriptor, inverseVariableSupply, (MultipleDelegateList<Object>) combinedList,
+        return new FlipSublistAction(listVariableDescriptor, (MultipleDelegateList<Object>) combinedList,
                 firstEndpoint, secondEndpoint);
     }
 

--- a/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java
+++ b/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java
@@ -9,7 +9,6 @@ import ai.timefold.solver.core.api.domain.solution.PlanningSolution;
 import ai.timefold.solver.core.api.score.director.ScoreDirector;
 import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescriptor;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
-import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonListInverseVariableDemand;
 import ai.timefold.solver.core.impl.heuristic.move.AbstractMove;
 import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
 import ai.timefold.solver.core.impl.util.Pair;
@@ -58,6 +57,34 @@ final class KOptListMove<Solution_> extends AbstractMove<Solution_> {
         for (int i = 0; i < originalEntities.length; i++) {
             originalEntities[i] = inverseVariableSupply.getInverseSingleton(combinedList.delegates[i].get(0));
         }
+    }
+
+    public KOptListMove(ListVariableDescriptor<Solution_> listVariableDescriptor,
+            KOptDescriptor<?> descriptor,
+            List<FlipSublistAction> equivalent2Opts,
+            MultipleDelegateList<?> combinedList,
+            int postShiftAmount,
+            int[] newEndIndices,
+            Object[] originalEntities) {
+        this.listVariableDescriptor = listVariableDescriptor;
+        this.descriptor = descriptor;
+        this.equivalent2Opts = equivalent2Opts;
+        this.postShiftAmount = postShiftAmount;
+        this.newEndIndices = newEndIndices;
+        this.combinedList = combinedList;
+        if (equivalent2Opts.isEmpty()) {
+            affectedElementsInfo = KOptAffectedElements.forMiddleRange(0, 0);
+        } else if (postShiftAmount != 0) {
+            affectedElementsInfo = KOptAffectedElements.forMiddleRange(0, equivalent2Opts.get(0).getCombinedList().size());
+        } else {
+            KOptAffectedElements currentAffectedElements = equivalent2Opts.get(0).getAffectedElements();
+            for (int i = 1; i < equivalent2Opts.size(); i++) {
+                currentAffectedElements = currentAffectedElements.merge(equivalent2Opts.get(i).getAffectedElements());
+            }
+            affectedElementsInfo = currentAffectedElements;
+        }
+
+        this.originalEntities = originalEntities;
     }
 
     KOptDescriptor<?> getDescriptor() {
@@ -117,14 +144,21 @@ final class KOptListMove<Solution_> extends AbstractMove<Solution_> {
     public KOptListMove<Solution_> rebase(ScoreDirector<Solution_> destinationScoreDirector) {
         List<FlipSublistAction> rebasedEquivalent2Opts = new ArrayList<>(equivalent2Opts.size());
         InnerScoreDirector<?, ?> innerScoreDirector = (InnerScoreDirector<?, ?>) destinationScoreDirector;
+        Object[] newEntities = new Object[originalEntities.length];
+        @SuppressWarnings("unchecked")
+        List<Object>[] newDelegates = new List[originalEntities.length];
 
+        for (int i = 0; i < newEntities.length; i++) {
+            newEntities[i] = innerScoreDirector.lookUpWorkingObject(originalEntities[i]);
+            newDelegates[i] = listVariableDescriptor.getListVariable(newEntities[i]);
+        }
+        MultipleDelegateList<Object> rebasedList = new MultipleDelegateList<>(newDelegates);
         for (FlipSublistAction twoOpt : equivalent2Opts) {
-            rebasedEquivalent2Opts.add(twoOpt.rebase(innerScoreDirector));
+            rebasedEquivalent2Opts.add(twoOpt.rebase(rebasedList));
         }
 
         return new KOptListMove<>(listVariableDescriptor,
-                innerScoreDirector.getSupplyManager().demand(new SingletonListInverseVariableDemand<>(listVariableDescriptor)),
-                descriptor, rebasedEquivalent2Opts, postShiftAmount, newEndIndices);
+                descriptor, rebasedEquivalent2Opts, rebasedList, postShiftAmount, newEndIndices, newEntities);
     }
 
     @Override

--- a/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java
+++ b/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java
@@ -11,7 +11,6 @@ import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescr
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonListInverseVariableDemand;
 import ai.timefold.solver.core.impl.heuristic.move.AbstractMove;
-import ai.timefold.solver.core.impl.heuristic.move.Move;
 import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
 import ai.timefold.solver.core.impl.util.Pair;
 
@@ -115,7 +114,7 @@ final class KOptListMove<Solution_> extends AbstractMove<Solution_> {
     }
 
     @Override
-    public Move<Solution_> rebase(ScoreDirector<Solution_> destinationScoreDirector) {
+    public KOptListMove<Solution_> rebase(ScoreDirector<Solution_> destinationScoreDirector) {
         List<FlipSublistAction> rebasedEquivalent2Opts = new ArrayList<>(equivalent2Opts.size());
         InnerScoreDirector<?, ?> innerScoreDirector = (InnerScoreDirector<?, ?>) destinationScoreDirector;
 

--- a/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateList.java
+++ b/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateList.java
@@ -11,7 +11,6 @@ import java.util.RandomAccess;
 import java.util.stream.Stream;
 
 import ai.timefold.solver.core.api.function.TriConsumer;
-import ai.timefold.solver.core.api.score.director.ScoreDirector;
 import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescriptor;
 import ai.timefold.solver.core.impl.domain.variable.index.IndexVariableSupply;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
@@ -27,7 +26,6 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
     final int[] delegateSizes;
     final int[] offsets;
     final int totalSize;
-    MultipleDelegateList<T> cachedRebaseResult;
 
     @SafeVarargs
     public MultipleDelegateList(List<T>... delegates) {
@@ -43,7 +41,6 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
         }
 
         this.totalSize = sizeSoFar;
-        this.cachedRebaseResult = null;
     }
 
     public int getIndexOfValue(ListVariableDescriptor<?> listVariableDescriptor,
@@ -57,25 +54,6 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
             }
         }
         throw new IllegalArgumentException("Value (" + value + ") is not contained in any entity list");
-    }
-
-    public MultipleDelegateList<T> rebase(ListVariableDescriptor<?> variableDescriptor,
-            SingletonInverseVariableSupply inverseVariableSupply, ScoreDirector<?> destinationScoreDirector) {
-        if (cachedRebaseResult != null) {
-            return cachedRebaseResult;
-        }
-        @SuppressWarnings("unchecked")
-        List<T>[] out = new List[delegates.length];
-        for (int i = 0; i < delegates.length; i++) {
-            List<T> delegate = delegates[i];
-            @SuppressWarnings("unchecked")
-            List<T> rebasedDelegate = (List<T>) variableDescriptor.getListVariable(
-                    destinationScoreDirector.lookUpWorkingObject(
-                            inverseVariableSupply.getInverseSingleton(delegate.get(0))));
-            out[i] = rebasedDelegate;
-        }
-        cachedRebaseResult = new MultipleDelegateList<>(out);
-        return cachedRebaseResult;
     }
 
     public void actOnAffectedElements(Object[] originalEntities, TriConsumer<Object, Integer, Integer> action) {

--- a/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateList.java
+++ b/core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateList.java
@@ -27,6 +27,7 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
     final int[] delegateSizes;
     final int[] offsets;
     final int totalSize;
+    MultipleDelegateList<T> cachedRebaseResult;
 
     @SafeVarargs
     public MultipleDelegateList(List<T>... delegates) {
@@ -42,6 +43,7 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
         }
 
         this.totalSize = sizeSoFar;
+        this.cachedRebaseResult = null;
     }
 
     public int getIndexOfValue(ListVariableDescriptor<?> listVariableDescriptor,
@@ -59,6 +61,9 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
 
     public MultipleDelegateList<T> rebase(ListVariableDescriptor<?> variableDescriptor,
             SingletonInverseVariableSupply inverseVariableSupply, ScoreDirector<?> destinationScoreDirector) {
+        if (cachedRebaseResult != null) {
+            return cachedRebaseResult;
+        }
         @SuppressWarnings("unchecked")
         List<T>[] out = new List[delegates.length];
         for (int i = 0; i < delegates.length; i++) {
@@ -69,7 +74,8 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
                             inverseVariableSupply.getInverseSingleton(delegate.get(0))));
             out[i] = rebasedDelegate;
         }
-        return new MultipleDelegateList<>(out);
+        cachedRebaseResult = new MultipleDelegateList<>(out);
+        return cachedRebaseResult;
     }
 
     public void actOnAffectedElements(Object[] originalEntities, TriConsumer<Object, Integer, Integer> action) {
@@ -241,7 +247,7 @@ public final class MultipleDelegateList<T> implements List<T>, RandomAccess {
     }
 
     @Override
-    public List<T> subList(int startInclusive, int endExclusive) {
+    public MultipleDelegateList<T> subList(int startInclusive, int endExclusive) {
         if (startInclusive < 0) {
             throw new IndexOutOfBoundsException("Sublist start index (" + startInclusive + ") out of range");
         }

--- a/core/core-impl/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMoveTest.java
+++ b/core/core-impl/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMoveTest.java
@@ -3,9 +3,11 @@ package ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt;
 import static ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.KOptUtils.getMultiEntityBetweenPredicate;
 import static ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.KOptUtils.getMultiEntitySuccessorFunction;
 import static ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.KOptUtils.getSuccessorFunction;
+import static ai.timefold.solver.core.impl.testdata.util.PlannerTestUtils.mockRebasingScoreDirector;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.reset;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 import java.util.List;
 import java.util.function.Function;
@@ -15,6 +17,7 @@ import ai.timefold.solver.core.impl.domain.variable.index.IndexVariableDemand;
 import ai.timefold.solver.core.impl.domain.variable.index.IndexVariableSupply;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonListInverseVariableDemand;
+import ai.timefold.solver.core.impl.domain.variable.supply.SupplyManager;
 import ai.timefold.solver.core.impl.heuristic.move.AbstractMove;
 import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
 import ai.timefold.solver.core.impl.testdata.domain.list.TestdataListEntity;
@@ -23,6 +26,7 @@ import ai.timefold.solver.core.impl.testdata.domain.list.TestdataListValue;
 import ai.timefold.solver.core.impl.testdata.util.PlannerTestUtils;
 
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
 
 class KOptListMoveTest {
 
@@ -44,6 +48,19 @@ class KOptListMoveTest {
     private final TestdataListValue v10 = new TestdataListValue("10");
     private final TestdataListValue v11 = new TestdataListValue("11");
     private final TestdataListValue v12 = new TestdataListValue("12");
+
+    private final TestdataListValue destinationV1 = new TestdataListValue("1");
+    private final TestdataListValue destinationV2 = new TestdataListValue("2");
+    private final TestdataListValue destinationV3 = new TestdataListValue("3");
+    private final TestdataListValue destinationV4 = new TestdataListValue("4");
+    private final TestdataListValue destinationV5 = new TestdataListValue("5");
+    private final TestdataListValue destinationV6 = new TestdataListValue("6");
+    private final TestdataListValue destinationV7 = new TestdataListValue("7");
+    private final TestdataListValue destinationV8 = new TestdataListValue("8");
+    private final TestdataListValue destinationV9 = new TestdataListValue("9");
+    private final TestdataListValue destinationV10 = new TestdataListValue("10");
+    private final TestdataListValue destinationV11 = new TestdataListValue("11");
+    private final TestdataListValue destinationV12 = new TestdataListValue("12");
 
     // TODO: It appears the multi-entity approach does not like kopt-affected-elements;
     // (in particular, I found index variable corruption causing a NPE due to incorrect after
@@ -77,6 +94,112 @@ class KOptListMoveTest {
         assertThat(e1.getValueList()).containsExactly(v1, v2, v3, v4, v5, v6);
         verify(scoreDirector).beforeListVariableChanged(variableDescriptor, e1, 0, 6);
         verify(scoreDirector).afterListVariableChanged(variableDescriptor, e1, 0, 6);
+    }
+
+    @Test
+    void test3OptRebase() {
+        TestdataListEntity e1 = TestdataListEntity.createWithValues("e1", v1, v2, v3, v4, v5, v6);
+        TestdataListEntity destinationE1 = TestdataListEntity.createWithValues("e1", destinationV1, destinationV2,
+                destinationV3, destinationV4, destinationV5, destinationV6);
+
+        KOptListMove<TestdataListSolution> kOptListMove = fromRemovedAndAddedEdges(scoreDirector,
+                variableDescriptor,
+                List.of(v6, v1,
+                        v2, v3,
+                        v4, v5),
+                List.of(v1, v3,
+                        v2, v5,
+                        v4, v6));
+
+        InnerScoreDirector<TestdataListSolution, ?> destinationScoreDirector = mockRebasingScoreDirector(
+                variableDescriptor.getEntityDescriptor().getSolutionDescriptor(), new Object[][] {
+                        { v1, destinationV1 },
+                        { v2, destinationV2 },
+                        { v3, destinationV3 },
+                        { v4, destinationV4 },
+                        { v5, destinationV5 },
+                        { v6, destinationV6 },
+                        { e1, destinationE1 },
+                });
+        SupplyManager supplyManager = Mockito.mock(SupplyManager.class);
+        SingletonInverseVariableSupply inverseVariableSupply = Mockito.mock(SingletonInverseVariableSupply.class);
+
+        when(destinationScoreDirector.getSupplyManager()).thenReturn(supplyManager);
+        when(supplyManager.demand(Mockito.any(SingletonListInverseVariableDemand.class))).thenReturn(inverseVariableSupply);
+        when(inverseVariableSupply.getInverseSingleton(destinationE1.getValueList().get(0))).thenReturn(destinationE1);
+
+        KOptListMove<TestdataListSolution> rebasedMove = kOptListMove.rebase(destinationScoreDirector);
+        assertThat(rebasedMove.isMoveDoable(destinationScoreDirector)).isTrue();
+        AbstractMove<TestdataListSolution> undoMove = rebasedMove.doMove(destinationScoreDirector);
+        assertThat(destinationE1.getValueList()).containsExactly(destinationV1, destinationV2, destinationV5, destinationV6,
+                destinationV4, destinationV3);
+        verify(destinationScoreDirector).beforeListVariableChanged(variableDescriptor, destinationE1, 0, 6);
+        verify(destinationScoreDirector).afterListVariableChanged(variableDescriptor, destinationE1, 0, 6);
+
+        reset(destinationScoreDirector);
+        undoMove.doMoveOnly(destinationScoreDirector);
+        assertThat(destinationE1.getValueList()).containsExactly(destinationV1, destinationV2, destinationV3, destinationV4,
+                destinationV5, destinationV6);
+        verify(destinationScoreDirector).beforeListVariableChanged(variableDescriptor, destinationE1, 0, 6);
+        verify(destinationScoreDirector).afterListVariableChanged(variableDescriptor, destinationE1, 0, 6);
+    }
+
+    @Test
+    void testMultiEntity3OptRebase() {
+        TestdataListEntity e1 = TestdataListEntity.createWithValues("e1", v1, v2, v3, v6);
+        TestdataListEntity e2 = TestdataListEntity.createWithValues("e2", v4, v5);
+
+        TestdataListEntity destinationE1 =
+                TestdataListEntity.createWithValues("e1", destinationV1, destinationV2, destinationV3, destinationV6);
+        TestdataListEntity destinationE2 = TestdataListEntity.createWithValues("e2", destinationV4, destinationV5);
+
+        KOptListMove<TestdataListSolution> kOptListMove = fromRemovedAndAddedEdges(scoreDirector,
+                variableDescriptor,
+                List.of(v6, v1,
+                        v2, v3,
+                        v4, v5),
+                List.of(v1, v3,
+                        v2, v5,
+                        v4, v6));
+
+        InnerScoreDirector<TestdataListSolution, ?> destinationScoreDirector = mockRebasingScoreDirector(
+                variableDescriptor.getEntityDescriptor().getSolutionDescriptor(), new Object[][] {
+                        { v1, destinationV1 },
+                        { v2, destinationV2 },
+                        { v3, destinationV3 },
+                        { v4, destinationV4 },
+                        { v5, destinationV5 },
+                        { v6, destinationV6 },
+                        { e1, destinationE1 },
+                        { e2, destinationE2 },
+                });
+        SupplyManager supplyManager = Mockito.mock(SupplyManager.class);
+        SingletonInverseVariableSupply inverseVariableSupply = Mockito.mock(SingletonInverseVariableSupply.class);
+
+        when(destinationScoreDirector.getSupplyManager()).thenReturn(supplyManager);
+        when(supplyManager.demand(Mockito.any(SingletonListInverseVariableDemand.class))).thenReturn(inverseVariableSupply);
+        when(inverseVariableSupply.getInverseSingleton(destinationE1.getValueList().get(0))).thenReturn(destinationE1);
+        when(inverseVariableSupply.getInverseSingleton(destinationE2.getValueList().get(0))).thenReturn(destinationE2);
+
+        KOptListMove<TestdataListSolution> rebasedMove = kOptListMove.rebase(destinationScoreDirector);
+
+        assertThat(rebasedMove.isMoveDoable(destinationScoreDirector)).isTrue();
+        AbstractMove<TestdataListSolution> undoMove = rebasedMove.doMove(destinationScoreDirector);
+        assertThat(destinationE1.getValueList()).containsExactly(destinationV1, destinationV5);
+        assertThat(destinationE2.getValueList()).containsExactly(destinationV2, destinationV6, destinationV4, destinationV3);
+        verify(destinationScoreDirector).beforeListVariableChanged(variableDescriptor, destinationE1, 0, 4);
+        verify(destinationScoreDirector).beforeListVariableChanged(variableDescriptor, destinationE2, 0, 2);
+        verify(destinationScoreDirector).afterListVariableChanged(variableDescriptor, destinationE1, 0, 2);
+        verify(destinationScoreDirector).afterListVariableChanged(variableDescriptor, destinationE2, 0, 4);
+
+        reset(destinationScoreDirector);
+        undoMove.doMoveOnly(destinationScoreDirector);
+        assertThat(destinationE1.getValueList()).containsExactly(destinationV1, destinationV2, destinationV3, destinationV6);
+        assertThat(destinationE2.getValueList()).containsExactly(destinationV4, destinationV5);
+        verify(destinationScoreDirector).beforeListVariableChanged(variableDescriptor, destinationE1, 0, 2);
+        verify(destinationScoreDirector).beforeListVariableChanged(variableDescriptor, destinationE2, 0, 4);
+        verify(destinationScoreDirector).afterListVariableChanged(variableDescriptor, destinationE1, 0, 4);
+        verify(destinationScoreDirector).afterListVariableChanged(variableDescriptor, destinationE2, 0, 2);
     }
 
     @Test

--- a/core/core-impl/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateListTest.java
+++ b/core/core-impl/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateListTest.java
@@ -11,7 +11,6 @@ import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
-import ai.timefold.solver.core.api.score.director.ScoreDirector;
 import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescriptor;
 import ai.timefold.solver.core.impl.domain.variable.index.IndexVariableSupply;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
@@ -81,108 +80,6 @@ public class MultipleDelegateListTest {
                 "g"))
                 .isInstanceOf(IllegalArgumentException.class)
                 .hasMessage("Value (g) is not contained in any entity list");
-    }
-
-    @Test
-    void testRebase() {
-        List<Object> delegate1 = new ArrayList<>(List.of("a", "b", "c"));
-        List<Object> delegate2 = new ArrayList<>(List.of("d", "e"));
-        List<Object> delegate3 = new ArrayList<>(List.of("f"));
-        MultipleDelegateList<Object> combined = new MultipleDelegateList<>(delegate1, delegate2, delegate3);
-
-        List<Object> rebasedDelegate1 = new ArrayList<>(delegate1);
-        List<Object> rebasedDelegate2 = new ArrayList<>(delegate2);
-        List<Object> rebasedDelegate3 = new ArrayList<>(delegate3);
-
-        ListVariableDescriptor<?> listVariableDescriptor = Mockito.mock(ListVariableDescriptor.class);
-        Mockito.when(listVariableDescriptor.getListVariable("e1")).thenReturn(delegate1);
-        Mockito.when(listVariableDescriptor.getListVariable("e2")).thenReturn(delegate2);
-        Mockito.when(listVariableDescriptor.getListVariable("e3")).thenReturn(delegate3);
-        Mockito.when(listVariableDescriptor.getListVariable("rebasedE1")).thenReturn(rebasedDelegate1);
-        Mockito.when(listVariableDescriptor.getListVariable("rebasedE2")).thenReturn(rebasedDelegate2);
-        Mockito.when(listVariableDescriptor.getListVariable("rebasedE3")).thenReturn(rebasedDelegate3);
-
-        ScoreDirector<?> destinationScoreDirector = Mockito.mock(ScoreDirector.class);
-        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e1")).thenReturn("rebasedE1");
-        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e2")).thenReturn("rebasedE2");
-        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e3")).thenReturn("rebasedE3");
-
-        SingletonInverseVariableSupply inverseVariableSupply = object -> {
-            String value = (String) object;
-            return switch (value) {
-                case "a", "b", "c" -> "e1";
-                case "d", "e" -> "e2";
-                case "f" -> "e3";
-                default -> null;
-            };
-        };
-
-        MultipleDelegateList<Object> rebasedCombined = combined.rebase(listVariableDescriptor,
-                inverseVariableSupply,
-                destinationScoreDirector);
-
-        assertThat(combined.delegates).containsExactly(delegate1, delegate2, delegate3);
-        assertThat(rebasedCombined.delegates).containsExactly(rebasedDelegate1, rebasedDelegate2, rebasedDelegate3);
-
-        combined.set(0, "a2");
-        assertThat(delegate1.get(0)).isEqualTo("a2");
-        assertThat(rebasedDelegate1.get(0)).isEqualTo("a");
-
-        combined.set(5, "f2");
-        assertThat(delegate3.get(0)).isEqualTo("f2");
-        assertThat(rebasedDelegate3.get(0)).isEqualTo("f");
-
-        rebasedCombined.set(0, "a3");
-        assertThat(delegate1.get(0)).isEqualTo("a2");
-        assertThat(rebasedDelegate1.get(0)).isEqualTo("a3");
-
-        rebasedCombined.set(5, "f3");
-        assertThat(delegate3.get(0)).isEqualTo("f2");
-        assertThat(rebasedDelegate3.get(0)).isEqualTo("f3");
-    }
-
-    @Test
-    void testMultipleRebaseReturnSameResult() {
-        List<Object> delegate1 = new ArrayList<>(List.of("a", "b", "c"));
-        List<Object> delegate2 = new ArrayList<>(List.of("d", "e"));
-        List<Object> delegate3 = new ArrayList<>(List.of("f"));
-        MultipleDelegateList<Object> combined = new MultipleDelegateList<>(delegate1, delegate2, delegate3);
-
-        List<Object> rebasedDelegate1 = new ArrayList<>(delegate1);
-        List<Object> rebasedDelegate2 = new ArrayList<>(delegate2);
-        List<Object> rebasedDelegate3 = new ArrayList<>(delegate3);
-
-        ListVariableDescriptor<?> listVariableDescriptor = Mockito.mock(ListVariableDescriptor.class);
-        Mockito.when(listVariableDescriptor.getListVariable("e1")).thenReturn(delegate1);
-        Mockito.when(listVariableDescriptor.getListVariable("e2")).thenReturn(delegate2);
-        Mockito.when(listVariableDescriptor.getListVariable("e3")).thenReturn(delegate3);
-        Mockito.when(listVariableDescriptor.getListVariable("rebasedE1")).thenReturn(rebasedDelegate1);
-        Mockito.when(listVariableDescriptor.getListVariable("rebasedE2")).thenReturn(rebasedDelegate2);
-        Mockito.when(listVariableDescriptor.getListVariable("rebasedE3")).thenReturn(rebasedDelegate3);
-
-        ScoreDirector<?> destinationScoreDirector = Mockito.mock(ScoreDirector.class);
-        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e1")).thenReturn("rebasedE1");
-        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e2")).thenReturn("rebasedE2");
-        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e3")).thenReturn("rebasedE3");
-
-        SingletonInverseVariableSupply inverseVariableSupply = object -> {
-            String value = (String) object;
-            return switch (value) {
-                case "a", "b", "c" -> "e1";
-                case "d", "e" -> "e2";
-                case "f" -> "e3";
-                default -> null;
-            };
-        };
-
-        MultipleDelegateList<Object> rebasedCombined1 = combined.rebase(listVariableDescriptor,
-                inverseVariableSupply,
-                destinationScoreDirector);
-        MultipleDelegateList<Object> rebasedCombined2 = combined.rebase(listVariableDescriptor,
-                inverseVariableSupply,
-                destinationScoreDirector);
-
-        assertThat(rebasedCombined1).isSameAs(rebasedCombined2);
     }
 
     @Test

--- a/core/core-impl/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateListTest.java
+++ b/core/core-impl/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/MultipleDelegateListTest.java
@@ -109,19 +109,12 @@ public class MultipleDelegateListTest {
 
         SingletonInverseVariableSupply inverseVariableSupply = object -> {
             String value = (String) object;
-            switch (value) {
-                case "a":
-                case "b":
-                case "c":
-                    return "e1";
-                case "d":
-                case "e":
-                    return "e2";
-                case "f":
-                    return "e3";
-                default:
-                    return null;
-            }
+            return switch (value) {
+                case "a", "b", "c" -> "e1";
+                case "d", "e" -> "e2";
+                case "f" -> "e3";
+                default -> null;
+            };
         };
 
         MultipleDelegateList<Object> rebasedCombined = combined.rebase(listVariableDescriptor,
@@ -135,9 +128,61 @@ public class MultipleDelegateListTest {
         assertThat(delegate1.get(0)).isEqualTo("a2");
         assertThat(rebasedDelegate1.get(0)).isEqualTo("a");
 
+        combined.set(5, "f2");
+        assertThat(delegate3.get(0)).isEqualTo("f2");
+        assertThat(rebasedDelegate3.get(0)).isEqualTo("f");
+
         rebasedCombined.set(0, "a3");
         assertThat(delegate1.get(0)).isEqualTo("a2");
         assertThat(rebasedDelegate1.get(0)).isEqualTo("a3");
+
+        rebasedCombined.set(5, "f3");
+        assertThat(delegate3.get(0)).isEqualTo("f2");
+        assertThat(rebasedDelegate3.get(0)).isEqualTo("f3");
+    }
+
+    @Test
+    void testMultipleRebaseReturnSameResult() {
+        List<Object> delegate1 = new ArrayList<>(List.of("a", "b", "c"));
+        List<Object> delegate2 = new ArrayList<>(List.of("d", "e"));
+        List<Object> delegate3 = new ArrayList<>(List.of("f"));
+        MultipleDelegateList<Object> combined = new MultipleDelegateList<>(delegate1, delegate2, delegate3);
+
+        List<Object> rebasedDelegate1 = new ArrayList<>(delegate1);
+        List<Object> rebasedDelegate2 = new ArrayList<>(delegate2);
+        List<Object> rebasedDelegate3 = new ArrayList<>(delegate3);
+
+        ListVariableDescriptor<?> listVariableDescriptor = Mockito.mock(ListVariableDescriptor.class);
+        Mockito.when(listVariableDescriptor.getListVariable("e1")).thenReturn(delegate1);
+        Mockito.when(listVariableDescriptor.getListVariable("e2")).thenReturn(delegate2);
+        Mockito.when(listVariableDescriptor.getListVariable("e3")).thenReturn(delegate3);
+        Mockito.when(listVariableDescriptor.getListVariable("rebasedE1")).thenReturn(rebasedDelegate1);
+        Mockito.when(listVariableDescriptor.getListVariable("rebasedE2")).thenReturn(rebasedDelegate2);
+        Mockito.when(listVariableDescriptor.getListVariable("rebasedE3")).thenReturn(rebasedDelegate3);
+
+        ScoreDirector<?> destinationScoreDirector = Mockito.mock(ScoreDirector.class);
+        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e1")).thenReturn("rebasedE1");
+        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e2")).thenReturn("rebasedE2");
+        Mockito.when(destinationScoreDirector.lookUpWorkingObject("e3")).thenReturn("rebasedE3");
+
+        SingletonInverseVariableSupply inverseVariableSupply = object -> {
+            String value = (String) object;
+            return switch (value) {
+                case "a", "b", "c" -> "e1";
+                case "d", "e" -> "e2";
+                case "f" -> "e3";
+                default -> null;
+            };
+        };
+
+        MultipleDelegateList<Object> rebasedCombined1 = combined.rebase(listVariableDescriptor,
+                inverseVariableSupply,
+                destinationScoreDirector);
+        MultipleDelegateList<Object> rebasedCombined2 = combined.rebase(listVariableDescriptor,
+                inverseVariableSupply,
+                destinationScoreDirector);
+
+        assertThat(rebasedCombined1).isSameAs(rebasedCombined2);
     }
 
     @Test

